src/OrchardCore/OrchardCore.DisplayManagement/Notify/NotifyEntry.cs

```diff
@@ -12,7 +15,40 @@ public enum NotifyType
 
     public class NotifyEntry
     {
+        private (HtmlEncoder HtmlEncoder, string Message) _cache;
+
         public NotifyType Type { get; set; }
         public IHtmlContent Message { get; set; }
+
+        public string ToHtmlString(HtmlEncoder htmlEncoder)
+        {
+            // When the object is created from a cookie the message
+            // is an HtmlString so we can use this instead of using
+            // the TextWriter path.
+
+            if (Message is IHtmlString htmlString)
+            {
+                return htmlString.ToHtmlString();
+            }
+
+            // Cache the encoded version for the specified encoder.
+            // This is necessary as long as there will be string-based comparisons
+            // and the need of NotifyEntryComparer
+
+            var cache = _cache;
```

Comment:
I guess nothing here prevents from Message changing and cache being invalid, private setter for properties (or being get-only) would make a different in API promises